source/tools/MatchBin.h

```diff
@@ -0,0 +1,287 @@
+ *  @date 2019
+ *
+ *  @file MatchBin.h
+ *  @brief A container that supports flexible tag-based lookup. .
+ *
+ */
+
+
+#ifndef EMP_MATCH_BIN_H
+#define EMP_MATCH_BIN_H
+
+#include <iostream>
+#include <unordered_map>
+#include <unordered_set>
+#include <functional>
+#include <algorithm>
+#include <stdexcept>
+#include <limits>
+
+#include "../base/assert.h"
+#include "../base/vector.h"
+#include "../tools/IndexMap.h"
+
+namespace emp {
+  /// Metric for MatchBin stored in the struct so we can template on it
+  /// Returns the number of bits not in common between two BitSets
+  template<size_t width>
+  struct HammingDistance{
+    double operator()(const BitSet<width>& a, const BitSet<width>& b) const{
+      return (double)(a^b).CountOnes();
+    }
+  };
+
+  /// Metric gives the absolute difference between two integers
+  struct Difference{
+    double operator()(const int a, const int b) const {
+      return (double)abs(a-b);
+    }
+  };
+
+  /// Metric gives the matchings by the closest tag on or above itself. (Wraps)
+  template<int max_value = 1000>
+  struct Push{
+    double operator()(const size_t a, const size_t b) const {
+      size_t difference = ((max_value + 1) + b - a) % (max_value+1);
+      return (double)(difference % (max_value+1));
+    }
+  };
+
+  // Matches based on the longest segment of equal and uneqal bits in two bitsets
+  template<size_t width>
+  struct DowningStreak{//TODO The regulator may be a little bias for this.
+    double operator()(const emp::BitSet<width>& a, const emp::BitSet<width>& b){
+      auto bs = a^b;
+      size_t same = (~bs).LongestSegmentOnes();
+      size_t different = bs.LongestSegmentOnes();
+      double ps = ProbabilityKBitSequence(same);
+      double pd = ProbabilityKBitSequence(different);
+
+      // A result nearing 1 is a better match but the threshold picks lower first.
+      // So we must subtract from 1.0 to get the inverse.
+      return 1.0 - (pd / (ps + pd));
+    }
+
+    inline double ProbabilityKBitSequence(size_t k){
+      return (width - k + 1) / pow(2, k);
+    }
+  };
+  
+  template<size_t width>
+    struct DowningInteger{
+      double operator()(const emp::BitSet<width>& a, const emp::BitSet<width>& b){
+        emp::BitSet<width> bitDifference = ( a > b ? a - b : b - a);
+        unsigned int fields = bitDifference.GetFields();
+        double result = 0;
+        for (unsigned int i = 0; i < fields; ++i){
+          result += bitDifference.GetUInt(i) * pow(2, 32 * i);
+        }
+
+      return result;
+      }
+    };
+  /// Selector for MatchBin stored in the struct so we cn template on it
+  /// returns sorted matches that are inside the given threshold 
+  template<typename ratio>
+  struct ThreshSelector{
+    emp::vector<size_t> operator()(emp::vector<size_t>& uids, std::unordered_map<size_t, double>& scores, size_t n){
+        
+      double thresh = ratio::num / ratio::den;
+      unsigned int i = 0;
+      if (n < log2(uids.size())){
```

Comment:
Where are log2 and pow coming from?? We didn’t #include “math.h” in this file I wonder if it’s somehow getting included via catch. Let’s #include <cmath> and then prefix these with std::